homeassistant/components/mqtt/__init__.py

```diff
@@ -432,23 +424,29 @@ def async_publish_service(call):
     return True
 
 
+Subscription = namedtuple('Subscription',
+                          ['topic', 'callback', 'qos', 'encoding'])
+Subscription.__new__.__defaults__ = (0, 'utf-8')
+
+Message = namedtuple('Message', ['topic', 'payload', 'qos', 'retain'])
+Message.__new__.__defaults__ = (0, False)
+
+
 class MQTT(object):
     """Home Assistant MQTT client."""
 
     def __init__(self, hass, broker, port, client_id, keepalive, username,
                  password, certificate, client_key, client_cert,
-                 tls_insecure, protocol, will_message, birth_message,
-                 tls_version):
+                 tls_insecure, protocol, will_message: Optional[Message],
+                 birth_message: Optional[Message], tls_version):
         """Initialize Home Assistant MQTT client."""
         import paho.mqtt.client as mqtt
 
         self.hass = hass
         self.broker = broker
         self.port = port
         self.keepalive = keepalive
-        self.wanted_topics = {}
-        self.subscribed_topics = {}
-        self.progress = {}
+        self.subscriptions = []
```

Comment:
This list now holds all active subscriptions with the `Subscription` tuple defined before. When a new subscription arrives it's first added to this list and then the paho `subscribe` call is made.
On a reconnect, this is also used to re-create all subscriptions as before.